flume-ng-core/src/main/java/org/apache/flume/source/ExecSource.java

```diff
@@ -165,30 +165,27 @@
 
   @Override
   public void start() {
-    logger.info("Exec source starting with command:{}", command);
+    logger.info("Exec source starting with command: {}", command);
 
-    executor = Executors.newSingleThreadExecutor();
+    // Start the counter before starting any threads that may access it.
+    sourceCounter.start();
 
-    runner = new ExecRunnable(shell, command, getChannelProcessor(), sourceCounter,
-        restart, restartThrottle, logStderr, bufferCount, batchTimeout, charset);
+    executor = Executors.newSingleThreadExecutor();
+    runner = new ExecRunnable(shell, command, getChannelProcessor(), sourceCounter, restart,
+                              restartThrottle, logStderr, bufferCount, batchTimeout, charset);
 
-    // FIXME: Use a callback-like executor / future to signal us upon failure.
```

Comment:
I don't really understand the purpose of this, actually I don't think it makes any sense because the startup is async and start() only fails synchronously. There isn't really another mechanism that I know of to "inform the rest of flume that ExecSource stopped unexpectedly". Or is there, and I just don't know about it?